### PR TITLE
Don't use explicit sync fences if we can tear

### PR DESCRIPTION
**Pull request by ferrreo**
#### Describe your PR, what does it fix/add?
Fixes https://github.com/hyprwm/Hyprland/issues/8482

This fixes the issue where with explicit sync on no tearing is possible even when the requirements for it are fully met.

This is similar to how KWIN handles it.

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
I am not 100% sure this is ultimately the correct fix but it retains fences and allows tearing to happen when requested. It also seems to be what other compositors are doing too.

Also been running this patch on my distro with a good number of users with no issues reported for around 3 weeks.

#### Is it ready for merging, or does it need work?
Yea do it.

### CODE REVIEW

File: src/render/Renderer.cpp [hunk 1]
```diff
@@ -1531,7 +1531,7 @@ bool CHyprRenderer::commitPendingAndDoExplicitSync(PHLMONITOR pMonitor) {
     if (inFD >= 0)
         pMonitor->output->state->setExplicitInFence(inFD);
     auto explicitOptions = getExplicitSyncSettings();
-    if (explicitOptions.explicitEnabled && explicitOptions.explicitKMSEnabled)
+    if (explicitOptions.explicitEnabled && explicitOptions.explicitKMSEnabled && !pMonitor->tearingState.canTear)
```
Comment (vaxerski) [inline src/render/Renderer.cpp:1534]: this should be activelyTearing not canTear.

**Review by vaxerski — COMMENTED** [review]
can I get a ref on kwin for this one?

**Comment by vaxerski** [comment]
yeah looks like it. we release when backend releases regardless

File: src/render/Renderer.cpp (hunk 1 above)
Comment (ferrreo) [reply src/render/Renderer.cpp:1534]: I tried that first and it didn't work

**Comment by ferrreo** [comment]
> can I get a ref on kwin for this one?

https://invent.kde.org/plasma/kwin/-/merge_requests/4800/diffs#f4ee7be3c0eb6f483546d5b8823ff33266747680_76_75

**Comment by vaxerski** [comment]
1) that's IN_FENCE? And you are disabling OUT_FENCE?

**Comment by UjinT34** [comment]
What does `explicitOutFence` really fences and when we should wait for it?

> This Sync File contains the CRTC fence that will be signaled when all framebuffers present on the Atomic Commit * request for that given CRTC are scanned out on the screen.

https://drmdb.emersion.fr/properties/3435973836/OUT_FENCE_PTR

Seems reasonable not to use it with tearing. 

kwin doesn't use `OUT_FENCE_PTR` at all and disables `IN_FENCE_FD` to avoid some bugs.

**Comment by vaxerski** [comment]
we could disable IN_FENCE, but I don't think that's a good idea: explicit sync + scanout without IN_FENCE _will_ mess up.

**Comment by UjinT34** [comment]
Is it possible to get DS with tearing?
https://github.com/hyprwm/Hyprland/blob/f9e4998a6d7713b19947b0db2c43ad88c5b71d80/src/render/Renderer.cpp#L1315

**Comment by vaxerski** [comment]
hm, in our code, maybe not, but ideally yea it should be perfectly possible

File: src/render/Renderer.cpp (hunk 1 above)
Comment (vaxerski) [reply src/render/Renderer.cpp:1534]: canTear is definitely wrong, but I don't thing we even use the out fence, can we drop it?

**Comment by vaxerski** [comment]
[patch.txt](https://github.com/user-attachments/files/18028375/patch.txt)
can't we just do this? (path for main). This just drops out fd, I don't think we actually use it at all.

**Comment by ferrreo** [comment]
Doesn't that patch undo https://github.com/hyprwm/Hyprland/commit/967fe76a60ec6b2928b495ed202f772beca05275?

**Comment by vaxerski** [comment]
the cursor does look choppy when tearing, but oh well.

**Comment by UjinT34** [comment]
With explicit sync enabled HL expects to receive the fence from AQ and complains
`Aquamarine did not return an explicit out fence`

I don't know how this fence should be used. Looks like it should be waited either before committing the next frame or before reusing the current framebuffer. Waiting for it immediately after the commit feels too early. Not using it at all is probably the same as waiting for it before reusing the fb because it is very unlikely to block at that point. Maybe there is no point in this fence with double buffering.

**Comment by vaxerski** [comment]
yeah I don't think there is.

**Comment by vaxerski** [comment]
We can go with my patch, do you want me to commit it to main or will you apply it here and we roll with this MR?

**Comment by ferrreo** [comment]
> We can go with my patch, do you want me to commit it to main or will you apply it here and we roll with this MR?

I think Just stick yer patch in main and I'll close this

**Pull request closed without merging**